[WIP] Automated retrieval of PDB

steven-albanese opened the pull request:
I've gotten started on this. Right now it's SUPER simple and you can only search by inhibitor. It's also kind of hacky and not very pretty at the moment.

jchodera commented:
The main objective right now is to automate the collection of a bunch of kinase-inhibitor PDB structure organized by kinase-inhibitor pair for approved target interactions, like:
```
pdbs/Imatinib-Abl/2HYY.pdb
                            /YYYY.pdb
    /Ponatinib-Abl/XXXX.pdb
...
We can certainly add other modes as we go along, but can you make it do something like this?

steven-albanese commented:
As I'm already using PDBfixer to retrieve the pdbs, would there be interest in having an option to 'fix' the PDB files? It could include any or all of the following: 
add missing residues, add hydrogens, replace non-standard AA, remove chains after 1

I'm mostly done so this shouldn't be too much extra to implement

steven-albanese commented:
I just updated this to work with that directory scheme and to by default search for ligand/approved target pairs . As a word of caution, the `csv` file isn't complete for somethings like sorafenib, which has a ton of targets.

steven-albanese commented:
TO DO: 

- Update CSV file to include all approved targets (Necessary) 
- add functionality to search for non-human kinase structures by using the UniProt name (if wanted) 
- add PDBfixer clean up mode to fix PDBs as they're downloaded (if wanted)

jchodera commented:
The option to fix the PDBs would be great. We would want to keep the ligand and optionally crystallographic waters.

Later we can add something to replace the ligand with the fully protonated form. I bet if we know the ligand PDB residue name we can just grab the fully protonated form from the LigandExpo Chemical Components Dictionary, but that might best be done as a feature to PDBFixer.

jchodera commented:
Actually, can we just run "protonate" on the resulting structure? I think that does the same thing (retrieving from the Ligand Expo)

jchodera commented:
@steven-albanese : Did you somehow mess up the carriage returns in the CSV file?  What text editor did you use?

steven-albanese commented:
What? I was using emacs

jchodera commented on goodiebag/approved/clinical-kinase-inhibitors.csv:
```diff
@@ -1,33 +1 @@
Palbociclib,Cc1c2cnc(nc2n(c(=O)c1C(=O)C)C3CCCC3)Nc4ccc(cn4)N5CCNCC5,CDK4 CDK6,CDK4 CDK6 CDK2 CDK1 CDK5 EGFR FGFR PDGFR InsR VEGFR AMPK CHK1 CK1 CK2 CSK GSK3 MAPK2 MAPKAPK2 PKB PKC PHK SAPK2 SAPK3 SAPK4,LQQ,P11802 Q00534
Cobimetinib,C1CCN[C@@H](C1)C2(CN(C2)C(=O)C3=C(C(=C(C=C3)F)F)NC4=C(C=C(C=C4)I)F)O,MEK,MEK1 MAPK2,EUI,Q02750
Osimertinib,CN1C=C(C2=CC=CC=C21)C3=NC(=NC=C3)NC4=C(C=C(C(=C4)NC(=O)C=C)N(C)CCN(C)C)OC,EGFR,EGFR,YY3,P00533
Alectinib,CCC1=C(C=C2C(=C1)C(=O)C3=C(C2(C)C)NC4=C3C=CC(=C4)C#N)N5CCC(CC5)N6CCOCC6,ALK,ALK INSR VEGFR2 ABL EGFR FGFR2 HER2 IGF1R JAK1 KIT MET PDGFR SRC AKT1 AKT2 AKT3 AuroraA CDK1 CDK2 MEK1 PKA PKC RAF,EMH,Q9UM73
```
See how there is only one `+` sign?  GitHub thinks you deleted everything and added back one long line because of some carriage return issue.

jchodera commented:
For some reason, the carriage returns are all messed up.  Not critical now---it turns out there is a "universal-newline" mode you can open files in, which works now.

jchodera commented:
But if you `cat clinical-kinase-inhibitors.csv` you can see it looks like it is all on one line.

jchodera commented:
Why do some of the compounds have multiple entries under `Chem_ID`?

Bosutinib seems to have both [`DB8`](http://ligand-expo.rcsb.org/reports/D/DB8/DB8_D3L0.gif) and [`XZN`](http://ligand-expo.rcsb.org/reports/X/XZN/XZN_D3L0.gif), but if you look these up in the ligand expo, they are different chemical structures:
![db8](https://cloud.githubusercontent.com/assets/3656088/14939939/ffecce22-0f2b-11e6-80ed-85d325111eb5.gif)
![xzn](https://cloud.githubusercontent.com/assets/3656088/14939940/02cc9960-0f2c-11e6-9167-a1b5eb4b0b77.gif)

steven-albanese commented:
Some have multiple names. Bosutinib has two for the two different isomers. I can remove them if wanted

jchodera commented:
We probably only want the FDA-approved isomer!

steven-albanese commented:
Ok just fixed the bosutinib isomer issue and added some PDBfixer functionality. There's now a flag (`--fix`) that will fix common mistakes and add hydrogens (`--ph` to specify, default at 7.0). Also a flag to remove heterogens (`--apo` will remove ligand AND crystal waters. To keep crystal waters use '--apo' and '--cwater' but remove the ligand).

steven-albanese commented:
I'm digging through the other comments that I've gotten tagged in, it seems like there's a `protonate` function that might allow me to protonate the ligand?

jchodera commented:
It's actually called [`reduce`](https://pypi.anaconda.org/openeye).  It's redistributed along with AmberTools, and while it isn't yet in `ambermini`, I think you can get it through this channel:

https://anaconda.org/mw/reduce

jchodera commented:
> Ok just fixed the bosutinib isomer issue and added some PDBfixer functionality

Are there other molecules with multiple entries?  Afatinib shows `0WN OWM`.  Searching the [Ligand Expo](http://ligand-expo.rcsb.org/ld-search.html) for `0WN` brings up something, but `OWM` returns no hits.

jchodera commented:
[`KIM`](http://ligand-expo.rcsb.org/reports/K/KIM/KIM_D3L0.gif) and [`SAV`](http://ligand-expo.rcsb.org/reports/S/SAV/SAV_D3L0.gif) are two different compounds that differ by the methyl on the nitrogen.  Only one of those is Pazopanib.  Is the other one a metabolite, or a derivative that is not Pazopanib?

jchodera commented:
Wikipedia says [the methylated one](https://en.wikipedia.org/wiki/Pazopanib) is pazopanib (`KIM`).

[This paper](http://www.tandfonline.com/doi/abs/10.3109/00498254.2012.734642?journalCode=ixen20#.Vye6GGN59gY) might suggest that `SAV` is a metabolite, but I can't get access to it.

steven-albanese commented:
Made those changes. The last one with multiple entries is Crizotinib, I can look into which one is which, although on `VGH` actually returns anything

jchodera commented:
OK, so ditch `VGS`?

jchodera commented:
Eventually, we may want to compile another list of metabolites for kinase inhibitors...

jchodera commented:
Or at least the bioactive forms!

steven-albanese commented:
Just to make sure I'm not missing it, is there a python API for reduce?

jchodera commented:
I don't think so. I think it's a command-line tool.

steven-albanese commented:
Ok- Just having a bit of a hold up because PDBfixer is deleting the ligand everywhere for  some reason, as well as adding in some amino acids in weird places

steven-albanese commented:
Ok I see. Some of the pdb files have the PDB files have the chemicals stored in different chains, which means if I want to create PDB files with only one chain I probably will be deleting the inhibitor. Have we encountered this problem before?

jchodera commented:
I have.  I had opened [this PDBFixer issue](https://github.com/pandegroup/pdbfixer/issues/53) motivated by similar issues.  You can definitely tell PDBFixer you want to preserve certain residue span / chain combinations by iterating over the residues and marking things to delete, but I can't see some way to avoid this without doing some manual curation about what you want to preserve. 

What do you think the best approach is for building this set?  Manual curation of the chain/residue spans to preserve for each pdbid?  Or something automated based on keeping the residue closest to a given chain?

steven-albanese commented:
I would love to have an automated system for this, as that would help when looking beyond the inhibitor:approved target set (which is relatively small). But for now, it might be ok to curate the chains to keep for each of the PDBs

jchodera commented:
OK, can you tackle that, and we'll brainstorm a more automated way to do this for the [kinase inhibitor benchmark](https://github.com/choderalab/kinase-benchmark)

steven-albanese commented:
sounds good- I think I'll probably just make another `.csv` file and read it in as a dictionary like I do with the approved inhibitors file

steven-albanese commented:
Ok- I've got a curated list for the chains to remove from all of the PDBs that we need and I've implemented a way to remove those chains and leave the pdbs that already only consist of chains we're interested in. However, I've run into some trouble pdb fixing `4WKQ.pdb`: 

```
Traceback (most recent call last):
  File "query_pdb.py", line 293, in <module>
    pdb_fix(s, pathway, ph, curated_chains)
  File "query_pdb.py", line 168, in pdb_fix
    fixer.addMissingAtoms()
  File "/Users/albaness/miniconda3/lib/python3.5/site-packages/pdbfixer/pdbfixer.py", line 804, in addMissingAtoms
    system = forcefield.createSystem(newTopology)
  File "/Users/albaness/miniconda3/lib/python3.5/site-packages/simtk/openmm/app/forcefield.py", line 812, in createSystem
    mass = self._atomTypes[typename].mass
AttributeError: 'tuple' object has no attribute 'mass'
```
I'm not sure what is causing that error, but I can fig through `forcefield.py` some more tomorrow.

steven-albanese commented:
Oh, I've also implemented a query mode `LigAll` which will cycle through all of the known FDA-approved inhibitors, incase you want the full set. It takes a fair bit of time if using the `--fix` flag at the same time

steven-albanese commented:
Add, I figured it out. It was a nonstandard amino acid. I added the two lines of code to find them and replace them

jchodera commented:
Awesomeness!

steven-albanese commented:
I think this is ready to merge, but I do want to point out that some of the PDBs have ridiculous looking N and C-terminal 'tails' that were added in by PDBfixer. I'm not sure if that will be a problem for the mcce people

jchodera commented:
> I think this is ready to merge, but I do want to point out that some of the PDBs have ridiculous looking N and C-terminal 'tails' that were added in by PDBfixer. I'm not sure if that will be a problem for the mcce people

Is it possible to limit the modeled region to the sequence spanning the first and last residues in the X-ray structure?

steven-albanese commented:
It looks like `findMissingResidues` creates a dictionary (fixer.missingResidues) where each key is a tuple of the chain and index at which the missing AA residues should be added. It looks like it might be easiest to `fixer.missingResidues.pop((0,0))`. This should eliminate the N-terminal fragments, which are what I believe to be the offenders in this situation (although I could be wrong)

jchodera commented:
You want to make sure you eliminate both N- and C- terminal fragments as long as there are no non-missing residues before or after them!

steven-albanese commented:
ok, so the N-terminus is easy, as the only time you would want to remove something is if it's inserted at (0,0). I can get the last insertion by calling `sorted(fixer.missingResidues.keys())[-1]` which will give me the last key. `sorted(fixer.missingResidues.keys())[-1][-1]` with give me the position at which the last insertion is occurring, as therefore the last resolved residue. I'm still not sure how to check if there is a non-missing residue after that. Is there a way to use `fixer.topology` to access the number of residues in chain 1 only?

jchodera commented:
I think you can do something like (requires `natsort` installed via `conda` or `pip`):
```python
# Get min and max resid in first chain
from natsort import natsorted
chains = [ chain for chain in fixer.topology.chains() ]
resids = [ residue.id for residue in chains[0].residues() ]
resids = natsorted(resids)
first_resid = resids[0]
last_resid = resids[-1]
```

steven-albanese commented:
oh great, that would help a lot!

steven-albanese commented:
Looks like that did the trick. Just gonna test this out and and then regenerate the set of PDBs

jchodera commented:
Woot!

I'm actually kind of surprised that worked.

steven-albanese commented:
ok, finished regenerating the files. They should be ready to go for the mcce folks and anyone else who wants to use it. Can someone merge this? I think @juliebehr might be the only one who can do that

juliebehr commented:
Q: should the pdb files themselves be in the repo, or can we get away with just the scripts themselves to be run by a user? also, I have no idea what the function of .gitignore is?

jchodera commented:
Maybe we don't want the PDB files in `goodiebag`, but we can certainly put them in https://github.com/choderalab/mcce-charges

steven-albanese commented:
I removed the PDBs and added a script to use this on the cluster. The .gitignore was so that I didn't accidentally commit pdb files while I was testing the script out

juliebehr commented:
Cool beans!

jchodera commented:
Hm, we don't seem to have picked up anything with ponatinib:Abl despite the existence of [a crystal structure](http://www.rcsb.org/pdb/explore/explore.do?structureId=3OXZ).  This one is maybe important because it has the titratable piperazine ring in the opposite orientation from bosutinib.

I think this is because the CSV file lists `BCR-ABL` instead of `ABL` as the target.  Maybe we could add `ABL`?

steven-albanese commented:
That crystal structure isn't picked up because it's Mouse Abl, which doesn't get picked up by the UniProtKB accession ID. I currently don't use the 'target' column for anything other than printing information for the user. I can add the UniProt Gene name as another search mode, which would pick that up. In addition, it's also possible to get this structure by using these flags:  `-l Ponatinib --mode Lig`
